New Hardhat tutorial (#40)
ethereum-optimism/optimism-tutorial: hardhat/README.md

# Using Optimistic Ethereum with the Hardhat Development Environment

[![Discord](https://img.shields.io/discord/667044843901681675.svg?color=768AD4&label=discord&logo=https%3A%2F%2Fdiscordapp.com%2Fassets%2F8c9701b98ad4372b58f13fd9f65f966e.svg)](https://discord.com/channels/667044843901681675)
[![Twitter Follow](https://img.shields.io/twitter/follow/optimismPBC.svg?label=optimismPBC&style=social)](https://twitter.com/optimismPBC)

This tutorial aims to help you get started with developing decentralized applications on [Optimistic Ethereum](https://optimism.io/) using [Hardhat](https://hardhat.org/). Applications 
running on top of Optimistic Ethereum are about as secure as those running on the underlying Ethereum mainnet itself, but are
[significantly cheaper](https://optimism.io/gas-comparison).

## Build an Optimistic Ethereum Server

> To test and debug on Optimistic Ethereum you need to have a running Optimistic Ethereum server, so the first step is to build one. The directions in this section are 

Reviewer: Maybe something along the lines of "The easiest way to test and debug apps on Optimistic Ethereum is to run a local Optimistic Ethereum node."

Author: I agree, except I changed to "fastest". It is arguably easier to use Kovan.

for an Ubuntu 20.04 VM running on GCP with a 20 GB disk (the default, 10 GB, is not enough), but they should be similar for other Linux 
versions and other platforms

### Install Prerequisite Software

> 1. Install [Docker](https://www.docker.com/). If you prefer not to use the convenience script, 

Reviewer: ```suggestion
1. Install [Docker](https://www.docker.com/). If you prefer not to use the convenience script shown below, 
```

   [there are other installation methods](https://docs.docker.com/engine/install/ubuntu).

   ```sh
   curl -fsSL https://get.docker.com -o get-docker.sh
   sudo sh get-docker.sh
   ```

2. Configure Docker settings:

   ```sh
   sudo usermod -a -G docker `whoami`

>    sudo apt install -y docker-compose

Reviewer: Maybe have a separate section: "3. Install [Docker Compose](https://docs.docker.com/compose/install/)". Not strongly opinionated.

    ```

> 3. Install [Node.js](https://nodejs.org/en/) and a number of npm packages. The version in the OS repository is 

Reviewer: ```suggestion
3. Install [Node.js](https://nodejs.org/en/). The version in the OS repository is 
```

  out of date, so we'll get the package from a different source.

   ```sh
   curl -sL https://deb.nodesource.com/setup_12.x -o nodesource_setup.sh
   sudo bash nodesource_setup.sh
   sudo apt install -y nodejs
   ```
   

> 4. Install the Node.js packages we need.

Reviewer: Maybe can link to this guide for installing globally without sudo if the user wishes: https://github.com/sindresorhus/guides/blob/main/npm-global-without-sudo.md

   ```sh   
   sudo npm install -g yarn hardhat
   ```
   

> 5. Log out and log back in to refresh the group information.

Reviewer: ```suggestion
5. Log out and log back in to complete the Docker installation (required).
```

> ### Start an Optimistic Ethereum Server

Reviewer: ```suggestion
### Start an Optimistic Ethereum Node
```

> This process downloads, compiles, and builds an Optimistic Ethereum network. Note that it takes a long time.

Reviewer: ```suggestion
This process downloads, compiles, and builds an Optimistic Ethereum network. Note that this may take a long time (up to 10 minutes).
```

Author: It took me more than that, so I'm reluctant to specify an amount of time.

```sh

> git clone https://github.com/ethereum-optimism/optimism.git

Reviewer: Possibly worth separating out the `git clone` + `cd optimism` step into a new code block with a comment like "First, clone the [Optimism monorepo](https://github.com/ethereum-optimism/optimism)"

cd optimism
yarn install
yarn build
cd ops
export COMPOSE_DOCKER_CLI_BUILD=1
export DOCKER_BUILDKIT=1
docker-compose build # --parallel
```

> Note that you will see a **Done** message at some point during the build process. Ignore it,

Reviewer: Hmmm maybe we can avoid needing this paragraph by doing `docker-compose build && echo "Build complete"` and telling people to just wait for the "Build complete" message.

it means that a specific section is done, and even though you do not see progress at that 
moment the build is continuing.

> This process is time consuming. You can continue the tutorial for now, I will note when you

Reviewer: Hmm this paragraph is a little confusing. Which process are you referring to?

Author: The build process. Can you see if the replacement is clearer?

The build process is time consuming, and you do not need to wait for it to finish before you continue the tutorial.
I will note the point in the tutorial where you need to have a running Optimistic Ethereum Node. Hopefully it will
be finished by then (you will know when the build process is done because you'll see a **Build complete** message).
After that you just need to start the node using:

```sh
docker-compose up
```

When you start seeing log entries scrolling on the console it means the node is now running.

need to stop and wait for it to finish.

```sh

> docker-compose up

Reviewer: I would add a sentence or two above explaining what this command does.

Reviewer: Hmmm I assume this may be here erroneously

Author: Added.

```

> When start seeing log entries scrolling on the console it means the system is now running. 

Reviewer: You can also run `./ops/scripts/wait-for-sequencer.sh` to know when the Sequencer is ready to go.

## Migrate a Dapp to Optimistic Ethereum

Now that we have Optimistic Ethereum running, it is time to run a distributed application (dapp) on it.

> **Note:** If you don't need the explanations and just want to see running code, 

Reviewer: IMO the `hardhat/dapp` directory should have a README with only the basics (`npx hardhat test` and `npx hardhat test --network optimistic`)

Author: But because of the Solidity version issue that doesn't always work.

`hardhat/dapp` directory
is just an `npm install` away from being a working example.

### Get a Sample Application

The easiest way is to start with a sample application. 

1. Open a second command line terminal
2. Run `hardhat`, the development environment we use in this tutorial
   ```sh
   mkdir dapp
   cd dapp
   npx hardhat
   ```
3. Select **Create a sample project** and accept all the defaults.
4. Verify the sample application.
   ```sh
   npx hardhat test
   ```
   
#### Interact with the Sample App Manually (optional)   
   
If you want to be more hands on, you can interact with the contract manually.

1. Start the console
   ```sh
   npx hardhat console
   ```
2. Deploy the greeter contract.
   ```javascript
   const Greeter = await ethers.getContractFactory("Greeter")
   const greeter = await Greeter.deploy("Hello, world!")
   await greeter.deployed()
   ```
3. Get the current greeting.
   ```javascript
   await greeter.greet()
   ```
4. Modify the greeting.
   ```javascript
   const tx = await greeter.setGreeting("Hola, mundo")
   await tx.wait()
   ```
5. Verify the greeting got modified.
   ```javascript
   await greeter.greet()
   ```

>    

Reviewer: Note to self: we should add markdown linting/formatting to this repository.

6. Leave the console.
   ```javascript
   .exit
   ```

### Migrate the Sample App to Optimistic Ethereum

> Now that we have a running Optimistic Ethereum server, and a dapp to run on it, we can run on Optimistic Ethereum.

Reviewer: ```suggestion
Now that we have a running Optimistic Ethereum node and a dapp to run on it, we can deploy to Optimistic Ethereum.
```

> 1. Install the Optimistic Ethereum package in the application.

Reviewer: ```suggestion
1. Install the Optimistic Ethereum hardhat plugin.
```

   ```sh
   yarn add @eth-optimism/hardhat-ovm
   ```
2. Edit `hardhat.config.js` to use the Optimistic Ethereum package.
   ```js
   require("@nomiclabs/hardhat-waffle");
   require('@eth-optimism/hardhat-ovm')

   ...
   ```
3. In the same file, add `optimistic` to the list of networks:
   ```js
   // hardhat.config.js

   ...
   
   module.exports = {
     solidity: "0.8.4",   // or whatever you get from HardHat
     networks: {
       // Add this network to your config!
       optimistic: {
          url: 'http://127.0.0.1:8545',

>           accounts: {

Reviewer: Can remove the `accounts` field and `gasPrice: 0` because the default hardhat accounts are now funded automatically.

Author: Sorry, but that's not true. When I remove `gasPrice` I get this error:

```

  Greeter
    1) Should return the new greeting once it's changed

  0 passing (30ms)
  1 failing

  1) Greeter
       Should return the new greeting once it's changed:
     Error: failed to get consistent fee data (operation="signer.getFeeData", code=UNSUPPORTED_OPERATION, version=abstract-signer/5.4.0)
      at Logger.makeError (node_modules/@ethersproject/logger/src.ts/index.ts:213:28)
      at Logger.throwError (node_modules/@ethersproject/logger/src.ts/index.ts:225:20)
      at Wallet.<anonymous> (node_modules/@ethersproject/abstract-signer/src.ts/index.ts:278:28)
      at step (node_modules/@ethersproject/abstract-signer/lib/index.js:48:23)
      at Object.next (node_modules/@ethersproject/abstract-signer/lib/index.js:29:53)
      at fulfilled (node_modules/@ethersproject/abstract-signer/lib/index.js:20:58)
      at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

When I remove `accounts` I get this error:

```

  Greeter
    1) Should return the new greeting once it's changed

  0 passing (274ms)
  1 failing

  1) Greeter
       Should return the new greeting once it's changed:
     Error: processing response error (body="{\"jsonrpc\":\"2.0\",\"id\":48,\"error\":{\"code\":-32000,\"message\":\"OVM: Unsupported RPC Method\"}}\n", error={"code":-32000}, requestBody="{\"method\":\"eth_sendTransaction\",\"params\":[{\"gas\":\"0x374b4dd\",\"from\":\"0x7e5f4552091a69125d5dfcb7b8c2659029395bdf\",\"data\":\"0x60806040523480156200001c576000806200001962000330565b50505b50604051620010b3380380620010b3833981810160405260208110156200004d576000806200004a62000330565b50505b810190808051604051939291908464010000000082111562000079576000806200007662000330565b50505b838201915060208201858111156200009b576000806200009862000330565b50505b8251866001820283011164010000000082111715620000c457600080620000c162000330565b50505b8083526020830192505050908051906020019080838360005b83811015620000fa578082015181840152602081019050620000dd565b50505050905090810190601f168015620001285780820380516001836020036101000a031916815260200191505b506040525050506200015f6040518060600160405280602281526020016200109160229139826200017f60201b620003601760201c565b806000908051906020019062000177929190620003a0565b50506200066d565b620002f58282604051602401808060200180602001838103835285818151815260200191508051906020019080838360005b83811015620001ce578082015181840152602081019050620001b1565b50505050905090810190601f168015620001fc5780820380516001836020036101000a031916815260200191505b50838103825284818151815260200191508051906020019080838360005b83811015620002375780820151818401526020810190506200021a565b50505050905090810190601f168015620002655780820380516001836020036101000a031916815260200191505b509450505050506040516020818303038152906040527f4b5c4277000000000000000000000000000000000000000000000000000000007bffffffffffffffffffffffffffffffffffffffffffffffffffffffff19166020820180517bffffffffffffffffffffffffffffffffffffffffffffffffffffffff8381831617835250505050620002f960201b60201c565b5050565b60008151905060006a636f6e736f6c652e6c6f679050602083016000808483855a620003246200046c565b50505050505050505050565b632a2a7adb598160e01b8152600481016020815285602082015260005b868110156200036d5780860151816040840101526020810190506200034d565b506020828760640184336000905af158600e01573d6000803e3d6000fd5b3d6001141558600a015760016000f35b505050565b8280620003ac62000577565b600181600116156101000203166002900490600052602060002090601f016020900481019282620003ec57600085620003e4620005dc565b505062000459565b82601f106200041257805160ff191683800117856200040a620005dc565b505062000459565b8280016001018562000423620005dc565b5050821562000459579182015b82811115620004585782518262000446620005dc565b50509160200191906001019062000430565b5b50905062000468919062000643565b5090565b638540661f598160e01b8152620004ae565b60008190508183111562000490578290505b92915050565b600081905081831015620004a8578290505b92915050565b836004820152846024820152606060448201528660648201526084810160005b88811015620004eb578088015181830152602081019050620004ce565b506060828960a40184336000905af158600e01573d6000803e3d6000fd5b3d6001141558600a015760016000f35b815160408301513d6000853e8b8b82606087013350600060045af15059620005428d3d62000496565b8c016200055081876200047e565b5b828110156200056a576000815260208101905062000551565b50839d5050505050505050565b6303daa959598160e01b8152836004820152602081602483336000905af158600e01573d6000803e3d6000fd5b3d6001141558600a015760016000f35b8051935060005b6040811015620005d757600081830152602081019050620005bb565b505050565b6322bd64c0598160e01b8152836004820152846024820152600081604483336000905af158600e01573d6000803e3d6000fd5b3d6001141558600a015760016000f35b60005b60408110156200063e5760008183015260208101905062000622565b505050565b5b8082111562000669576000816000906200065d620005dc565b50505060010162000644565b5090565b610a14806200067d6000396000f3fe6080604052348015610019576000806100166106db565b50505b506004361061003f5760003560e01c8063a41368621461004d578063cfae32171461012c575b60008061004a6106db565b50505b61012a6004803603602081101561006c576000806100696106db565b50505b81019080803590602001906401000000008111156100925760008061008f6106db565b50505b8201836020820111156100ad576000806100aa6106db565b50505b803590602001918460018302840111640100000000831117156100d8576000806100d56106db565b50505b91908080601f016020809104026020016040519081016040528093929190818152602001838380828437600081840152601f19601f8201169050808301925050505050505091929192905050506101af565b005b6101346102a2565b6040518080602001828103825283818151815260200191508051906020019080838360005b83811015610174578082015181840152602081019050610159565b50505050905090810190601f1680156101a15780820380516001836020036101000a031916815260200191505b509250505060405180910390f35b6102886040518060600160405280602381526020016109f1602391396000806101d6610749565b600181600116156101000203166002900480601f01602080910402602001604051908101604052809291908181526020018280610211610749565b6001816001161561010002031660029004801561027d5780601f1061024b57610100808361023d610749565b04028352916020019161027d565b820191906000526020600020905b81610262610749565b8152906001019060200180831161025957829003601f168201915b5050505050836104cc565b806000908051906020019061029e9291906107ac565b5050565b60606000806102af610749565b600181600116156101000203166002900480601f016020809104026020016040519081016040528092919081815260200182806102ea610749565b600181600116156101000203166002900480156103565780601f10610324576101008083610316610749565b040283529160200191610356565b820191906000526020600020905b8161033b610749565b8152906001019060200180831161033257829003601f168201915b5050505050905090565b6104c88282604051602401808060200180602001838103835285818151815260200191508051906020019080838360005b838110156103ac578082015181840152602081019050610391565b50505050905090810190601f1680156103d95780820380516001836020036101000a031916815260200191505b50838103825284818151815260200191508051906020019080838360005b838110156104125780820151818401526020810190506103f7565b50505050905090810190601f16801561043f5780820380516001836020036101000a031916815260200191505b509450505050506040516020818303038152906040527f4b5c4277000000000000000000000000000000000000000000000000000000007bffffffffffffffffffffffffffffffffffffffffffffffffffffffff19166020820180517bffffffffffffffffffffffffffffffffffffffffffffffffffffffff83818316178352505050506106a6565b5050565b6106a183838360405160240180806020018060200180602001848103845287818151815260200191508051906020019080838360005b8381101561051d578082015181840152602081019050610502565b50505050905090810190601f16801561054a5780820380516001836020036101000a031916815260200191505b50848103835286818151815260200191508051906020019080838360005b83811015610583578082015181840152602081019050610568565b50505050905090810190601f1680156105b05780820380516001836020036101000a031916815260200191505b50848103825285818151815260200191508051906020019080838360005b838110156105e95780820151818401526020810190506105ce565b50505050905090810190601f1680156106165780820380516001836020036101000a031916815260200191505b5096505050505050506040516020818303038152906040527f2ced7cef000000000000000000000000000000000000000000000000000000007bffffffffffffffffffffffffffffffffffffffffffffffffffffffff19166020820180517bffffffffffffffffffffffffffffffffffffffffffffffffffffffff83818316178352505050506106a6565b505050565b60008151905060006a636f6e736f6c652e6c6f679050602083016000808483855a6106cf610865565b50505050505050505050565b632a2a7adb598160e01b8152600481016020815285602082015260005b868110156107165780860151816040840101526020810190506106f8565b506020828760640184336000905af158600e01573d6000803e3d6000fd5b3d6001141558600a015760016000f35b505050565b6303daa959598160e01b8152836004820152602081602483336000905af158600e01573d6000803e3d6000fd5b3d6001141558600a015760016000f35b8051935060005b60408110156107a75760008183015260208101905061078d565b505050565b82806107b6610749565b600181600116156101000203166002900490600052602060002090601f0160209004810192826107f2576000856107eb610965565b5050610854565b82601f1061081457805160ff1916838001178561080d610965565b5050610854565b82800160010185610823610965565b50508215610854579182015b8281111561085357825182610842610965565b50509160200191906001019061082f565b5b50905061086191906109ca565b5090565b638540661f598160e01b81526108a4565b600081905081831115610887578290505b92915050565b60008190508183101561089e578290505b92915050565b836004820152846024820152606060448201528660648201526084810160005b888110156108df5780880151818301526020810190506108c4565b506060828960a40184336000905af158600e01573d6000803e3d6000fd5b3d6001141558600a015760016000f35b815160408301513d6000853e8b8b82606087013350600060045af150596109348d3d61088d565b8c016109408187610876565b5b828110156109585760008152602081019050610941565b50839d5050505050505050565b6322bd64c0598160e01b8152836004820152846024820152600081604483336000905af158600e01573d6000803e3d6000fd5b3d6001141558600a015760016000f35b60005b60408110156109c5576000818301526020810190506109ab565b505050565b5b808211156109ec576000816000906109e1610965565b5050506001016109cb565b509056fe4368616e67696e67206772656574696e672066726f6d202725732720746f20272573274465706c6f79696e67206120477265657465722077697468206772656574696e673a0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000d48656c6c6f2c20776f726c642100000000000000000000000000000000000000\"}],\"id\":48,\"jsonrpc\":\"2.0\"}", requestMethod="POST", url="http://127.0.0.1:8545", code=SERVER_ERROR, version=web/5.4.0)
      at Logger.makeError (node_modules/@ethersproject/logger/src.ts/index.ts:213:28)
      at Logger.throwError (node_modules/@ethersproject/logger/src.ts/index.ts:225:20)
      at /home/ori/optimism-tutorial/hardhat/dapp/node_modules/@ethersproject/web/src.ts/index.ts:284:28
      at step (node_modules/@ethersproject/web/lib/index.js:33:23)
      at Object.next (node_modules/@ethersproject/web/lib/index.js:14:53)
      at fulfilled (node_modules/@ethersproject/web/lib/index.js:5:58)
      at processTicksAndRejections (internal/process/task_queues.js:97:5)

```

             mnemonic: 'test test test test test test test test test test test junk'
          },
          // This sets the gas price to 0 for all transactions on L2. We do this
          // because account balances are not automatically initiated with an ETH
          // balance (yet, sorry!).
          gasPrice: 0,
          ovm: true // This sets the network as using the ovm and ensure contract will be compiled against that.
       }
     }
   }
   ```

>    At this point you need to wait until the `docker-compose build` ends, if it hasn't yet, and then run

Reviewer: I think this should be its own bullet point.

   `cd ~/optimism/ops ; docker-compose up`.

4. Test the contract on Optimistic Ethereum. Hardhat will recognize it has not been compiled and compile it for you.

   ```sh
   npx hardhat --network optimistic test
   ```

5. If you want to interact with the app manually, use the console. You can use the same JavaScript commands
   to control it you used above.
   ```sh
   npx hardhat --network optimistic console
   ```
   
   

> #### Change the Solidity Version if Needed

Reviewer: Hmmm. Do current versions of hardhat always give 0.8.0+? If so, maybe worth explaining this as soon as the user runs `npx hardhat` instead of here since they might run into this bug and be confused when running the above commands.

Author: The same program is for 0.8.4 (IIRC). But the above commands will all work, this is where it becomes a problem. Also, doing it here lets me illustrate that when you change anything you need to test it first in Ethereum and only then in Optimistic Ethereum.

To run on Optimistic Ethereum a contract needs to be compiled with a variant Solidity compiler. Sometimes
the latest version of Solidity supported by Optimistic Ethereum is not the same as the version used by the
sample app in HardHat. When that is the case, the `npx hardhat --network optimistic test` command
fails with an error message similar to:

```
OVM Compiler Error (insert "// @unsupported: ovm" if you don't want this file to be compiled for the OVM):
 contracts/Greeter.sol:2:1: ParserError: Source file requires different compiler version (current compiler is 0.7.6) - note that nightly builds are considered to be strictly less than the released version
pragma solidity ^0.8.0;
^---------------------^

Error HH600: Compilation failed
```

To solve this problem:

1. Edit the `hardhat.config.js` file to change `module.exports.solidity` to the supported version.
2. Edit `contracts/Greeter.sol` to change the `pragma solidity` line to the supported version.
3. Check the application still works on normal Ethereum.
   ```sh
   npx hardhat test
   ```
4. Check the application works on Optimistic Ethereum.
   ```sh
   npx hardhat --network optimistic test
   ```

> ## How to Run Tests

Reviewer: Hmmm section title is a little confusing. Maybe something like "Best practices for running tests".

> As you may have noticed, in this tutorial we ran all the tests first on the HardHat EVM and only then on Optimistic Ethereum. This is

Reviewer: I think worth expanding on this section a little. Just needs 1-2 more sentences about why this is best practice. I can help if you'd like.

Author: I don't have anymore to say, so if you do I'd appreciate your addition.

# Using Optimistic Ethereum with the Hardhat Development Environment

[![Discord](https://img.shields.io/discord/667044843901681675.svg?color=768AD4&label=discord&logo=https%3A%2F%2Fdiscordapp.com%2Fassets%2F8c9701b98ad4372b58f13fd9f65f966e.svg)](https://discord.com/channels/667044843901681675)
[![Twitter Follow](https://img.shields.io/twitter/follow/optimismPBC.svg?label=optimismPBC&style=social)](https://twitter.com/optimismPBC)

This tutorial aims to help you get started with developing decentralized applications on [Optimistic Ethereum](https://optimism.io/) using [Hardhat](https://hardhat.org/). Applications 
running on top of Optimistic Ethereum are about as secure as those running on the underlying Ethereum mainnet itself, but are

> [significantly cheaper](https://optimism.io/gas-comparison).

Author: The whole section is copied, with one tiny change, to the Truffle tutorial. 

In general I'd rather have one copy of everything and use prerequisites to tell people if they need to do something prior to starting a tutorial. However, this is the initial time people use Optimism and they don't have much investment in the platform. Therefore, I want to make their lives as easy as possible.

> [significantly cheaper](https://optimism.io/gas-comparison).

Reviewer 2: So you think the tutorials readme page should just be empty and have links to hardhat, truffle, etc which have like 30% overlapping content?

> [significantly cheaper](https://optimism.io/gas-comparison).

Author: Do you mean the page at https://community.optimism.io/docs/resources/tutorials.html? I think it'll have copies of the tutorial readmes, automatically updated by a github action.

Hardhat and Truffle will have overlapping content, but the other tutorials, which go further in depth, will not.

> [significantly cheaper](https://optimism.io/gas-comparison).

Reviewer 2: I mean this https://github.com/ethereum-optimism/optimism-tutorial/blob/14cc6cac80c0702c551d409a9ae89defcaec7acb/README.md

> [significantly cheaper](https://optimism.io/gas-comparison).

Author: > I mean this https://github.com/ethereum-optimism/optimism-tutorial/blob/14cc6cac80c0702c551d409a9ae89defcaec7acb/README.md

Yes. I plan to put all the tutorials in a single repository, to keep the management relatively simple. So the top level README.md just be a menu of links to the various tutorials.

> ## Build an Optimistic Ethereum Node

Author: > There should be two options to run the optimism node (development network) - firstly building it from the repo - which this section describes, and the second should be to get a ready-made docker image with that environment already built and in place. (Not sure if we have such image yet though @snario might know?)

@smartcontracts  told me he'll try to figure out how to do it (I tried and failed). I'll add it to the tutorial when available, I just don't want to delay the initial release of the tutorial.
